repair/repair.cc

```diff
@@ -150,13 +150,104 @@ locator::node_set node_ops_cmd_request::get_leaving_nodes(const locator::topolog
     return ret;
 }
 
-std::unordered_map<gms::inet_address, gms::inet_address> node_ops_cmd_request::get_replace_nodes() {
-    if (nodes_dict.empty()) {
-        return std::move(replace_nodes);
+std::unordered_map<locator::node_ptr, locator::node_ptr> node_ops_cmd_request::get_replace_nodes(const locator::topology& topo) {
+    std::unordered_map<locator::node_ptr, locator::node_ptr> ret;
+    for (auto&& [x, y] : std::move(replace_nodes)) {
+        locator::host_id_and_endpoint replaced_haep, replacing_haep;
+        locator::node_ptr replaced_node, replacing_node;
+        if (!nodes_dict.empty()) {
+            replaced_haep = nodes_dict.at(x.raw_addr());
+            replaced_node = topo.find_node(replaced_haep.host_id);
+            replacing_haep = nodes_dict.at(y.raw_addr());
+            replacing_node = topo.find_node(replacing_haep.host_id);
+        } else {
+            replaced_haep.endpoint = x;
+            replaced_node = topo.find_node(replaced_haep.endpoint);
+            replacing_haep.endpoint = y;
+            replacing_node = topo.find_node(replacing_haep.endpoint);
+        }
+        if (!replaced_node) {
+            log_error_and_throw<std::runtime_error>(rlogger, "Replaced node {}/{} not found in topology", replaced_haep.host_id, replaced_haep.endpoint);
+        }
+        if (!replaced_node) {
+            log_error_and_throw<std::runtime_error>(rlogger, "Replacing node {}/{} not found in topology", replacing_haep.host_id, replacing_haep.endpoint);
+        }
+        ret.emplace(replaced_node, replacing_node);
     }
-    std::unordered_map<gms::inet_address, gms::inet_address> ret;
+    return ret;
+}
+
+static locator::node_ptr add_node(locator::topology& topo, const gms::gossiper& gossiper, const locator::host_id_and_endpoint& haep, locator::node::state state) {
+    auto* dc = gossiper.get_application_state_ptr(haep.endpoint, gms::application_state::DC);
+    auto* rack = gossiper.get_application_state_ptr(haep.endpoint, gms::application_state::RACK);
+    auto dc_rack = locator::endpoint_dc_rack{
+        .dc = dc ? dc->value : locator::endpoint_dc_rack::default_location.dc,
+        .rack = (dc && rack) ? rack->value : locator::endpoint_dc_rack::default_location.rack,
+    };
+    return topo.add_node(haep.host_id, haep.endpoint, dc_rack, state);
+}
```

Comment:
I'm confused: isn't topology the definitive source of node information? Why are we getting node data from gossip and adding it to topology?

Reply:
We may learn about new nodes from gossip.
This should be integrated in the future with raft topology changes.

Reply:
We learn about new nodes from node operations, where this function is called from.
gossip is only used here to get additional node attributed like dc and rack